crypto/slh_dsa/slh_dsa.c

```diff
@@ -0,0 +1,359 @@
+}
+
+/**
+ * @brief SLH-DSA Signature verification
+ * See FIPS 205 Section 9.3 Algorithm 20
+ *
+ * A signature consists of
+ *   r[n] random bytes
+ *   [k]*[1+a][n] FORS signature bytes
+ *   [h + d*len][n] Hyper tree signature bytes
+ *
+ * @param hctx Contains SLH_DSA algorithm functions and constants and the
+ *             public SLH_DSA key to use for verification.
+ * @param msg The message to verify. This may be encoded beforehand.
+ * @param msg_len The size of |msg|
+ * @param sig A signature to verify
+ * @param sig_len The size of |sig|
+ * @returns 1 if the signature verification succeeded or 0 otherwise.
+ */
+static int slh_verify_internal(SLH_DSA_HASH_CTX *hctx,
+                               const uint8_t *msg, size_t msg_len,
+                               const uint8_t *sig, size_t sig_len)
+{
+    const SLH_DSA_KEY *pub = hctx->key;
+    SLH_HASH_FUNC_DECLARE(pub, hashf);
+    SLH_ADRS_FUNC_DECLARE(pub, adrsf);
+    SLH_ADRS_DECLARE(adrs);
+    const SLH_DSA_PARAMS *params = pub->params;
+    uint32_t n = params->n;
+    const uint8_t *pk_seed, *pk_root; /* Pointers to elements in |pub| */
+    PACKET pkt, *sig_rpkt = &pkt; /* Points to the |sig| buffer */
+    uint8_t m_digest[SLH_MAX_M];
+    const uint8_t *md; /* This is a pointer into the buffer in m_digest_rpkt */
+    size_t md_len = MD_LEN(params); /* 21..40 bytes */
+    PACKET pkt2, *m_digest_rpkt = &pkt2; /* Points to m_digest buffer */
+    const uint8_t *r; /* Pointer to |sig_rpkt| buffer */
+    uint8_t pk_fors[SLH_MAX_N];
+    uint64_t tree_id;
+    uint32_t leaf_id;
+
+    /* Exit if public key is not set */
+    if (pub->pub == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_MISSING_KEY);
+        return 0;
+    }
+
+    /* Exit if signature is invalid size */
+    if (sig_len != params->sig_len
+            || !PACKET_buf_init(sig_rpkt, sig, sig_len))
+        return 0;
+    if (!PACKET_get_bytes(sig_rpkt, &r, n))
+        return 0;
+
+    adrsf->zero(adrs);
+
+    pk_seed = SLH_DSA_PK_SEED(pub);
+    pk_root = SLH_DSA_PK_ROOT(pub);
+
+    if (!hashf->H_MSG(hctx, r, pk_seed, pk_root, msg, msg_len,
+                      m_digest, sizeof(m_digest)))
+        return 0;
+
+    /*
+     * Get md (the first md_len bytes of m_digest to use in
+     * ossl_slh_fors_pk_from_sig(), and then retrieve the tree id and leaf id
+     * from the remaining bytes in m_digest.
+     */
+    if (!PACKET_buf_init(m_digest_rpkt, m_digest, sizeof(m_digest))
+            || !PACKET_get_bytes(m_digest_rpkt, &md, md_len)
+            || !get_tree_ids(m_digest_rpkt, params, &tree_id, &leaf_id))
+        return 0;
+
+    adrsf->set_tree_address(adrs, tree_id);
+    adrsf->set_type_and_clear(adrs, SLH_ADRS_TYPE_FORS_TREE);
+    adrsf->set_keypair_address(adrs, leaf_id);
+    return ossl_slh_fors_pk_from_sig(hctx, sig_rpkt, md, pk_seed, adrs,
+                                     pk_fors, sizeof(pk_fors))
+        && ossl_slh_ht_verify(hctx, pk_fors, sig_rpkt, pk_seed,
```

Comment:
Style wise you are using goto err and ret = ... in the sign function but are taking a different approach here.
It grates that they are different when they don't need to be style-wise and they are right next to each other.

Reply:
Code style is different because there is no allocations, and in that case it should return,,